tests/nexus_api_test.go

```diff
@@ -26,48 +26,187 @@ package tests
+									},
+								},
+							},
+						},
+					},
+				}, nil
+			},
+			assertion: func(res *nexus.ClientStartOperationResult[string], err error) {
+				var operationError *nexus.UnsuccessfulOperationError
+				s.Require().ErrorAs(err, &operationError)
+				s.Require().Equal(nexus.OperationStateFailed, operationError.State)
+				s.Require().Equal("deliberate test failure", operationError.Failure.Message)
+				s.Require().Equal(map[string]string{"k": "v"}, operationError.Failure.Metadata)
+				var details string
+				err = json.Unmarshal(operationError.Failure.Details, &details)
+				s.Require().NoError(err)
+				s.Require().Equal("details", details)
+			},
+		},
+		{
+			outcome: "handler_error",
+			handler: func(res *workflowservice.PollNexusTaskQueueResponse) (*nexuspb.Response, *nexuspb.HandlerError) {
+				return nil, &nexuspb.HandlerError{
+					ErrorType: string(nexus.HandlerErrorTypeInternal),
+					Failure:   &nexuspb.Failure{Message: "deliberate internal failure"},
+				}
+			},
+			assertion: func(res *nexus.ClientStartOperationResult[string], err error) {
+				var unexpectedError *nexus.UnexpectedResponseError
+				s.Require().ErrorAs(err, &unexpectedError)
+				// TODO: nexus should export this
+				s.Require().Equal(520, unexpectedError.Response.StatusCode)
+				s.Require().Equal("deliberate internal failure", unexpectedError.Failure.Message)
+			},
+		},
+		// TODO: This can't be tested without the test taking over a minute since this is the default matching
+		// client timeout and there's currently no way for the client to specify the request timeout.
+		// Tested manually for now.
+		// {
+		// 	outcome: "handler_timeout",
+		// 	handler: func(res *workflowservice.PollNexusTaskQueueResponse) (*nexuspb.Response, *nexuspb.HandlerError) {
+		// 		time.Sleep(time.Minute)
+		// 		return nil, nil
+		// 	},
+		// 	assertion: func(res *nexus.ClientStartOperationResult[string], err error) {
+		// 		var unexpectedError *nexus.UnexpectedResponseError
+		// 		s.Require().ErrorAs(err, &unexpectedError)
+		//              ...
+		// 	},
+		// },
+	}
 
-	s.Equal(1, len(snap["nexus_latency"]))
-	s.Equal(map[string]string{"namespace": s.namespace, "method": "StartOperation", "outcome": "internal_error"}, snap["nexus_latency"][0].Tags)
-	s.Equal(metrics.MetricUnit(metrics.Milliseconds), snap["nexus_latency"][0].Unit)
+	for _, tc := range testCases {
+		tc := tc
+		s.T().Run(tc.outcome, func(t *testing.T) {
+			taskQueue := s.randomizeStr("task-queue")
+			ctx := NewContext()
+
+			url := fmt.Sprintf("http://%s/api/v1/namespaces/%s/task-queues/%s/dispatch-nexus-task", s.httpAPIAddress, s.namespace, taskQueue)
+			client, err := nexus.NewClient(nexus.ClientOptions{ServiceBaseURL: url})
+			s.Require().NoError(err)
+			capture := s.testCluster.host.captureMetricsHandler.StartCapture()
+			defer s.testCluster.host.captureMetricsHandler.StopCapture(capture)
+
+			go s.nexusTaskPoller(ctx, taskQueue, tc.handler)
```

Comment:
I think it's worth adding a sync.WaitGroup here to be sure that this goroutine exits before the subtest completes.

Reply:
It's to ensure we aren't leaking goroutines. Reviewing `s.nexusTaskPoller`, it looks like there aren't, but it future-proofs us as well. It's a small thing, but I think it's worth it.